src/workerd/jsg/modules.c++

```diff
@@ -4,10 +4,38 @@
 
 #include "modules.h"
 #include "promise.h"
+#include <kj/mutex.h>
 
 namespace workerd::jsg {
 namespace {
 
+class CompileCache {
+  // The CompileCache is used to hold cached compilation data for built-in JavaScript modules.
+public:
+  void add(kj::StringPtr key, std::unique_ptr<v8::ScriptCompiler::CachedData> cached) const {
+    cache.lockExclusive()->insert(kj::str(key), kj::mv(cached));
+  }
+
+  kj::Maybe<v8::ScriptCompiler::CachedData&> find(kj::StringPtr key) const {
+    return cache.lockShared()->find(key).map([](auto& data)
+        -> v8::ScriptCompiler::CachedData& { return *data; });
+  }
+
+  static const CompileCache& get() {
+    static CompileCache instance;
+    return instance;
+  }
+
+private:
+  kj::MutexGuarded<kj::HashMap<kj::String, std::unique_ptr<v8::ScriptCompiler::CachedData>>> cache;
+  // TODO(review): There's a question here: Is it better to have a single CompileCache for the
```

Comment:
Since the cache is only locked briefly to perform an insert or lookup operation, and only when initially loading an isolate, contention will be negligible here, so it's much better to share the cache.